File: Sources/RemDesignSystem/Primitives/ContainedIconTokenSet.swift
import SwiftUI

/// The size axis of `ContainedIcon` — 1:1 with the Figma **Size** property on the
/// ContainedIcon variant set (`614:8`). Kept a public top-level enum (not nested in the
/// internal token set) so it can appear in `ContainedIcon`'s public API — the same shape as
/// `RemButtonSize`.
public enum ContainedIconSize: Sendable {
    case small   // 38pt — inline list-row leading
    case large   // 64pt — hero
}

/// Resolves every styleable value for a `ContainedIcon` `(fill, size)` into one value type —
/// the same **TokenSet** pattern as `RemButtonTokenSet`. Every value comes from `DesignTokens`
/// (one documented exception, below), so `ContainedIcon` itself stays thin and no literals live
/// in the view. Figma canonical: ContainedIcon variant set `614:8` (property **Fill**:
/// Tinted/Subtle × **Size**: Small/Large).
///
/// There is deliberately **no** `ContainedIconStyle`: unlike `RemButton`, a `ContainedIcon` is a
/// decorative `View`, not a control, so it has no rest/pressed/disabled states and no SwiftUI
/// `ButtonStyle` hook. The design-system parallel for a plain view is *TokenSet + View*; the
/// parallel for a control is *TokenSet + Style*.
struct ContainedIconTokenSet {
    let dimension: CGFloat
    let cornerRadius: CGFloat
    let glyphPointSize: CGFloat
    let background: Color
    let foreground: Color

    init(fill: ContainedIcon.Fill, size: ContainedIconSize) {
        switch size {
        case .small:
            dimension = 38
            cornerRadius = DesignTokens.CornerRadius.small          // 8
        case .large:
            dimension = 64
            // 18 is the 64pt hero squircle radius the founder approved. It is intentionally OFF
            // the 8/12/16/24 scale — flagged for token reconciliation; centralized here so it
            // lives in exactly one place rather than at call sites.
            cornerRadius = 18
        }
        glyphPointSize = dimension * 0.46

        switch fill {
        case .tint(let color):
            background = color
            foreground = DesignTokens.Color.labelOnColor
        case .subtle:
            background = DesignTokens.Color.fillTertiary
            foreground = DesignTokens.Color.labelSecondary
        }
    }
}

File: compose/RemDesignSystem/primitives/ContainedIconTokenSet.kt
package com.rem.designsystem.primitives

import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.rem.designsystem.tokens.RemColors
import com.rem.designsystem.tokens.RemRadius

/**
 * Size axis of `ContainedIcon` — 1:1 with the Figma **Size** property (ContainedIcon set `614:8`).
 * Public top-level enum (not nested in the internal token set) so it can appear in `ContainedIcon`'s
 * public API — the Kotlin twin of Swift `ContainedIconSize`.
 */
enum class ContainedIconSize { Small, Large }

/** Fill axis — 1:1 with the Figma **Fill** property: `Tinted` (solid + on-color glyph) / `Subtle`. */
sealed interface ContainedIconFill {
    data class Tint(val color: Color) : ContainedIconFill
    data object Subtle : ContainedIconFill
}

/**
 * Resolves every styleable value for a `(fill, size)` into one value type — the **TokenSet** pattern,
 * the Kotlin twin of `ContainedIconTokenSet.swift`. Every value comes from the generated `RemTokens`
 * (one documented exception below), so the `ContainedIcon` composable stays thin and no literals live
 * in the view.
 *
 * There is deliberately **no `ContainedIconStyle`**: unlike `RemButton`, a `ContainedIcon` is a
 * decorative view, not a control — no rest/pressed/disabled. The design-system parallel for a plain
 * view is *TokenSet + View*; for a control it is *TokenSet + Style*.
 */
internal data class ContainedIconTokens(
    val dimension: Dp,
    val cornerRadius: Dp,
    val glyphSize: Dp,
    val background: Color,
    val foreground: Color,
)

@Composable
internal fun containedIconTokens(fill: ContainedIconFill, size: ContainedIconSize): ContainedIconTokens {
    val colors = RemColors.current
    val dimension: Dp = if (size == ContainedIconSize.Large) 64.dp else 38.dp
    // 18dp is the 64dp hero squircle radius the founder approved — intentionally OFF the
    // 8/12/16/24 scale, flagged for token reconciliation, centralized here (mirrors the Swift note).
    val cornerRadius: Dp = if (size == ContainedIconSize.Large) 18.dp else RemRadius.small
    val glyphSize: Dp = dimension * 0.46f
    return when (fill) {
        is ContainedIconFill.Tint -> ContainedIconTokens(
            dimension = dimension,
            cornerRadius = cornerRadius,
            glyphSize = glyphSize,
            background = fill.color,
            // on-color glyph. Swift uses a `labelOnColor` token; it isn't on main's tokens.json yet,
            // so white here — reconcile by adding `color.label.onColor` to tokens.json + the generator.
            foreground = Color.White,
        )
        ContainedIconFill.Subtle -> ContainedIconTokens(
            dimension = dimension,
            cornerRadius = cornerRadius,
            glyphSize = glyphSize,
            background = colors.fillTertiary,
            foreground = colors.labelSecondary,
        )
    }
}
